test/helpers/position_assertions.cc

```diff
@@ -1187,44 +1205,58 @@ void SymbolSearchAssertion::checkAllForQuery(std::string query,
+        for (auto &assertion : assertionsOnLine) {
+            bool foundMatchingSymbol = false;
+            for (auto &symbol : symbolsOnLine) {
+                if (assertion->matches(uriPrefix, symbol)) {
+                    foundMatchingSymbol = true;
+                    break;
+                }
+            }
+            if (foundMatchingSymbol) {
+                continue;
+            }
+            numUnfulfilled++;
+            addFailureAtLocationWithSource(assertion->getLocation(uriPrefix), sourceFileContents, uriPrefix,
+                                           fmt::format("Unfulfilled assertion: {}", assertion->toString()));
         }
-        success = false;
-        auto assertion = expected[0];
-        auto location = assertion->getLocation(uriPrefix);
-        auto sourceLine = getLine(sourceFileContents, uriPrefix, *location.get());
-        ADD_FAILURE_AT(location->uri.c_str(), assertion->range->start->line + 1)
-            << fmt::format("WorkspaceSymbol query `{}` did not return a reference to:\n{}\n", query,
-                           prettyPrintRangeComment(sourceLine, *assertion->range, ""));
-    }
-
-    // Show any symbols returned by LSP that weren't asserted
-    for (auto entry : symbolsByLine) {
-        auto lineInfo = entry.first;
-        auto actual = entry.second;
-        auto &expected = assertionsByLine[lineInfo];
-        if (expected.size() == actual.size()) {
-            continue;
+        // Report symbols with no matching assertion
+        for (auto &symbol : symbolsOnLine) {
+            bool foundMatchingAssertion = false;
+            for (auto &assertion : assertionsOnLine) {
+                if (assertion->matches(uriPrefix, symbol)) {
+                    foundMatchingAssertion = true;
+                    break;
+                }
+            }
+            if (foundMatchingAssertion) {
+                continue;
+            }
+            numUnexpected++;
+            addFailureAtLocationWithSource(
+                symbol->location->copy(), sourceFileContents, uriPrefix,
+                fmt::format("Unexpected result from `workspace/symbol` query \"{}\": {}\n", query, symbol->toJSON()));
         }
-        success = false;
-        auto symbolInfo = actual[0];
-        auto &location = symbolInfo->location;
-        auto sourceLine = getLine(sourceFileContents, uriPrefix, *location.get());
-        ADD_FAILURE_AT(location->uri.c_str(), location->range->start->line + 1)
-            << fmt::format("WorkspaceSymbol query `{}` returned an unexpected reference to:\n{}\n", query,
-                           prettyPrintRangeComment(sourceLine, *location->range, ""));
     }
 
-    ASSERT_TRUE(success) << fmt::format("For query \"{}\"", query);
+    if (numUnfulfilled + numUnexpected > 0) {
+        ADD_FAILURE() << fmt::format(
+            "WorkspaceSymbol query \"{}\" failure{}: {} unfulfilled assertion{}, {} unexpected response{}\n", query,
```

Comment:
Maybe this is a better wording?

```suggestion
            "WorkspaceSymbol query \"{}\" failure{}: {} expected response{} missing, {} unexpected response{}\n", query,
```